AX: expose DOM identifiers via accessibilityIdentifier

hoffmanjoshua opened the pull request:
#### 80e2fadb90c3057600ad29d9605446f5bc6b7ddd
<pre>
AX: expose DOM identifiers via accessibilityIdentifier
<a href="https://bugs.webkit.org/show_bug.cgi?id=294462">https://bugs.webkit.org/show_bug.cgi?id=294462</a>
<a href="https://rdar.apple.com/153311467">rdar://153311467</a>

Reviewed by Andres Gonzalez.

When the `AXEnableWebKitDOMIdentifier` Accessibility user default is set to YES, allow
DOM identifiers to be returned via the accessibilityIdentifier method in the mac wrapper.

* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::AXObjectCache):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/ios/AXObjectCacheIOS.mm:
(WebCore::AXObjectCache::initializeUserDefaultValues):
* Source/WebCore/accessibility/mac/AXObjectCacheMac.mm:
(WebCore::AXObjectCache::initializeUserDefaultValues):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper accessibilityIdentifier]):

Canonical link: <a href="https://commits.webkit.org/296310@main">https://commits.webkit.org/296310@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3f5177474502b0d320e3a5550f930a4747069be4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/107933 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27598 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18017 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113144 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58455 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/109896 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28293 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36147 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/81923 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/110881 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22417 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97243 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62354 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/21829 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15376 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/57897 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/91768 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15442 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116271 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35005 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/25760 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/90956 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35381 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93521 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90750 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23171 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35655 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13408 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/30757 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/34903 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40457 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34646 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38006 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36307 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

lilyspiniolas commented on Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
```diff
@@ -2282,6 +2282,17 @@ - (NSArray *)accessibilityParameterizedAttributeNames
+    RetainPtr userDefaults = adoptNS([[NSUserDefaults alloc] initWithSuiteName:@"com.apple.Accessibility"]);
+    if ([userDefaults boolForKey:@"AXEnableWebKitDOMIdentifier"]) {
+        auto* backingObject = self.updateObjectBackingStore;
+        return backingObject ? backingObject->identifierAttribute().createNSString().autorelease() : nil;
```
```suggestion
        if (RefPtr backingObject = self.updateObjectBackingStore)
            return backingObject->identifierAttribute().createNSString().autorelease();
```

fleizach replied:
Does VoiceOver ever end up calling axIdentifier? If so this implementation might be pretty slow

hoffmanjoshua replied:
It looks like VO does occasionally call it. Maybe a better option here is to cache it once when the `AXObjectCache` is created, and the read that global value here.

AndresGonzalezApple replied:
The reason I think using the DOM identifiers as accessibilityIdentifiers  is not a good idea is because we promise in the API docs that it is a unique identifier. Many AX objects will have a "" identifier, not unique:
"Returns the unique ID for the accessibility element. It is often used in automated testing."
https://developer.apple.com/documentation/appkit/nsaccessibilityelementprotocol/accessibilityidentifier()?language=objc
If you cache the value of AXEnableWebKitDOMIdentifier, then you would need to register to observe changes (being rigorous).

hoffmanjoshua replied:
I think it's okay that many objects have a "" identifier. When this is used, for automation purposes, tests will be trying to target individual elements. Authors will make sure those elements they need have an identifier, so it's okay that others are empty.

The identifier API for native macOS UI elements works the same way. If no identifier is specified, it will be an empty string. Only once an author specifies the value does it return something unique.

hoffmanjoshua committed:
AX: expose DOM identifiers via accessibilityIdentifier
https://bugs.webkit.org/show_bug.cgi?id=294462
rdar://153311467

Reviewed by Andres Gonzalez.

When the `AXEnableWebKitDOMIdentifier` Accessibility user default is set to YES, allow
DOM identifiers to be returned via the accessibilityIdentifier method in the mac wrapper.

* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::AXObjectCache):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/ios/AXObjectCacheIOS.mm:
(WebCore::AXObjectCache::initializeUserDefaultValues):
* Source/WebCore/accessibility/mac/AXObjectCacheMac.mm:
(WebCore::AXObjectCache::initializeUserDefaultValues):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper accessibilityIdentifier]):

Canonical link: https://commits.webkit.org/296310@main

annevk commented on Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
```diff
@@ -2282,6 +2282,16 @@ - (NSArray *)accessibilityParameterizedAttributeNames
+- (NSString *)accessibilityIdentifier
+{
+    if (AXObjectCache::accessibilityDOMIdentifiersEnabled()) {
+        if (RefPtr<AXCoreObject> backingObject = self.updateObjectBackingStore)
```
What's the reason you went with `RefPtr<AXCoreObject>` rather than `RefPtr` here? The latter is preferred where possible.

hoffmanjoshua replied:
The compiler required a template parameter here, unfortunately. We also follow this same pattern throughout `WebAccessibilityObjectWrapperMac.mm`
